File: mvn-advent-of-code/src/main/java/advent/twenty_fifteen/Day25LetItSnow.java
package advent.twenty_fifteen;

import advent.common.DailyProblem;
import advent.utilities.FileUtilities;
import advent.utilities.Parsers;

public class Day25LetItSnow implements DailyProblem<Long, Long> {
    private final long part1Answer;
    private final long part2Answer;

    public Day25LetItSnow(String filename) {
        String input = FileUtilities.readLines(filename, Parsers::TO_STRING).get(0);
        String[] bits = input.split(" ");

        int targetRow = Integer.parseInt(bits[16].substring(0, bits[16].length() - 1)) - 1;
        int targetCol = Integer.parseInt(bits[18].substring(0, bits[18].length() - 1)) - 1;

        boolean finished = false;
        int r = 1;
        int c = 0;
        long previous = 20151125;
        while (!finished) {
            previous = (previous * 252533) % 33554393;
            r--;
            c++;
            if (r < 0) {
                r = c;
                c = 0;
            }
            finished = (r == targetRow) && (c == targetCol);
        }
        previous = (previous * 252533) % 33554393;

        this.part1Answer = previous;
        this.part2Answer = 0;
    }


    @Override
    public Long getPart1Answer() {
        return part1Answer;
    }

    @Override
    public Long getPart2Answer() {
        throw new UnsupportedOperationException("No Part 2");
    }
}

File: scala-advent-of-code/src/main/scala/advent/twenty_fifteen/Day25LetItSnow.scala
package advent.twenty_fifteen

import advent.common.DailyProblem
import advent.utilities.FileUtilities

import scala.annotation.tailrec

class Day25LetItSnow(filename: String) extends DailyProblem[Long, Void] {
  private val manualPattern = "To continue, please consult the code grid in the manual\\.  Enter the code at row ([0-9]+), column ([0-9]+)\\.".r

  private def parser(line: String): (Int, Int) =
    line match {
      case manualPattern(row, column) => (row.toInt, column.toInt)
      case _ => throw new IllegalStateException(s"Cannot match '$line'")
    }

  private val code = FileUtilities.readFile(filename, parser)(0)

  private def calc(previous : Long) : Long = (previous * 252533) % 33554393

  private def calculate(targetRow: Int, targetCol: Int): Long = {
    @tailrec
    def walk(r: Int, c: Int, previous: Long): Long = {
      if (r == targetRow && c == targetCol) {
        calc(previous)
      } else {
        if (r == 1) {
          walk(c + 1, 1,  calc(previous))
        } else {
          walk(r - 1, c + 1, calc(previous))
        }
      }
    }

    walk(2, 1, 20151125)
  }

  override val part1Answer: Long = calculate(code._1, code._2)
  override val part2Answer: Void = null
}
